store/store_download.go

```diff
@@ -343,15 +453,26 @@ func downloadImpl(ctx context.Context, name, sha3_384, downloadURL string, user
 		}
 		dlSize = float64(resp.ContentLength)
 		pbar.Start(name, dlSize)
-		mw := io.MultiWriter(w, h, pbar)
+		mw := io.MultiWriter(w, h, pbar, tc)
 		var limiter io.Reader
 		limiter = resp.Body
 		if limit := dlOpts.RateLimit; limit > 0 {
 			bucket := ratelimit.NewBucketWithRate(float64(limit), 2*limit)
 			limiter = ratelimitReader(resp.Body, bucket)
 		}
+
+		finish := tc.Monitor()
 		_, finalErr = io.Copy(mw, limiter)
+		close(finish)
 		pbar.Finished()
+
+		if cancelled(downloadCtx) {
+			if err := tc.Err(); err != nil {
```

Comment:
(nitpick) we could check tc.Err() even outside/before "if cancelled()", should do no harm and the fact that tc.Err is only set if there was a cancel is irrelevant.